### ISSUE TITLE
Request: exposing token utilization, getValue

### ISSUE THREAD

**Issue by amp127** [get-convex/ratelimiter#2]
This is a following conversation in the beta components thread. 

Context: Currently the api only provides you the next time that you can retry at. We reverse engineered the convex helper rate limiter to show a user their personal bucket status for that user account. Within this usecase sharding is not necessary avoiding the need to query a sample set. But I could see the use of that with a bigger token bucket with more resolution and activity. 

Ian suggested: 
```
const value = await rateLimiter.getValue(ctx, "myLimit");
// to just sample a few shards for efficiency / avoiding OCCs
const value = await rateLimiter.getValue(ctx, "myLimit", { sampleShards: 3 });
```
This works for us. But may have limited use for other users who don't want to reverse engineer it. 

I think it would be nice to have it optionally available when you call rateLimiter.check

**Comment by ianmacartney** [comment]
Hey sorry this fell through the cracks. Are you waiting for this before switching to the component? There's a few layers it could provide:

1. Expose the value, ts, and bucket metadata like window start time from the raw component API (you'd do `ctx.runQuery(components.rateLimiter.lib.getValue, { key: "myLimit" }` e.g.) or also in the "thick client" (rateLimiter.getValue)
2. Also expose a function that makes it easy to query what the value will be in the future, when it will reach a certain value, etc.
3. Make a convenience function to re-export the query, like `export const { getRateLimit } = rateLimiter.getter();` then in the browser have a react hook like `const { status, getValue, retryAt } = useRateLimit(api.foo.getRateLimit, "myLimit"[, 10]);` that does some math & clock skew calibration to let the client know when it could retry, or get the value at a certain time in the future, as well as stay subscribed to changes.

How much better is (3) than (1), and any other suggestions?

**Comment by amp127** [comment]
> Hey sorry this fell through the cracks. Are you waiting for this before switching to the component? There's a few layers it could provide:
> 
> 1. Expose the value, ts, and bucket metadata like window start time from the raw component API (you'd do `ctx.runQuery(components.rateLimiter.lib.getValue, { key: "myLimit" }` e.g.) or also in the "thick client" (rateLimiter.getValue)
> 2. Also expose a function that makes it easy to query what the value will be in the future, when it will reach a certain value, etc.
> 3. Make a convenience function to re-export the query, like `export const { getRateLimit } = rateLimiter.getter();` then in the browser have a react hook like `const { status, getValue, retryAt } = useRateLimit(api.foo.getRateLimit, "myLimit"[, 10]);` that does some math & clock skew calibration to let the client know when it could retry, or get the value at a certain time in the future, as well as stay subscribed to changes.
> 
> How much better is (3) than (1), and any other suggestions?

1 is needed for the widget as it stands. 

I would like to disable all buttons that are rate limited on the site, so i think 3 would be a little more helpful with that. 

The ultimate goal is to just have a component that renders a throttle usage that moves up and down and counts down automatically, so if we have to rewrite it a little to make it more accurate that is good.

**Comment by ianmacartney** [comment]
I published `@convex-dev/rate-limiter@0.2.8-alpha.0` which has a few nice things - copied from the readme:

### Using the React hook

You can use the React hook to check the rate limit in your browser code.

First, define the server API to get the rate limit value:

```ts
// In convex/example.ts
export const { getRateLimit, getServerTime } =
rateLimiter.hookAPI("sendMessage");
```

Then, use the React hook to check the rate limit:

```ts
function App() {
  const { status: { ok, retryAt }, check } =
useRateLimit(api.example.getRateLimit, {
    // All of these are optional
    count: 1, // The number of tokens to wait on
    getServerTimeMutation: getServerTime, // Allows the hook to align the
browser and server clocks
    key: userId, // The key to check for the rate limit, if you use one
  });

  // If you want to check at specific times and get the concrete value:
  const { value, ts, config, ok, retryAt } = check(Date.now(), count);
```

### Fetching the current value directly

You can fetch the current value of a rate limit directly, if you want to
know
the concrete value and timestamp it was last updated.

```ts
const { config, value, ts } = await rateLimiter.getValue(ctx,
"sendMessage", { key: userId });
```

And you can use `calculateRateLimit` to calculate the value at a given
timestamp:

```ts
import { calculateRateLimit } from ***@***.***/rate-limiter";

const { config, value, ts } = calculateRateLimit({ value, ts }, config,
Date.now(), count || 0);
```

This should suffice, but going to let it bake & use the alpha with the
Agent component before shipping it fully, in case there's some tweaks we
want

**Comment by amp127** [comment]
@ianmacartney 
Not sure if i'm missing something but i don't think the return validators are available in the public api?. 
No examples anywhere of using the validators .

**Comment by ianmacartney** [comment]
@amp127 sorry for the delay - want to put up a PR exposing the validators you need? My request would be that any validator exported is named something like `vRateLimitArgs`, `vFixedWindow` or `vConfig` instead of suffixed with Validator like `configValidator` as it is today
